### PR TITLE
[node] Add inspector type definitions

### PR DESCRIPTION
**Pull request by kjin**
This change adds type definitions for `inspector`, which was introduced in Node v8.

The `inspector` API is a little unique: an inspector `Session` object can issue commands through `post` and listen for events, and the list of possible commands and events are defined in the [2nd documentation link](https://chromedevtools.github.io/devtools-protocol/v8/) in the checklist below. Commands always take their inputs and outputs as single objects. Sessions emit their data as a single object per event as well.

Because there are a limited number of commands and events, and each command/event's input/output object shapes are known at compile-time, I've taken the liberty of defining interfaces capturing the structure of each command input, command output, and event output. Most of the type definitions (when counting lines) consist of these interface definitions.

With this, we can get code completion like in the following image (which I think is nice):

image

Most of the code is auto-generated from the [JSON file backing the 2nd documentation link](https://github.com/ChromeDevTools/devtools-protocol/blob/master/json/js_protocol.json). I believe the JSON file is auto-published as part of an npm package. (The auto-generation source code is located [here](https://github.com/DefinitelyTyped/DefinietlyTyped/types/node/scripts/generate-inspector).)

I think it's important to introduce at least some form of type definitions for `inspector`. However, I'm concerned that this change adds a disproportionally large amount of code (~35% delta) for a relatively seldom-used module in Node. That, in combination with its auto-generated content, would likely compound on the maintenance burden for Node type definitions. So I'd like to ask any potential reviewers to weigh in on this: what is the best way to introduce type definitions for `inspector`?

I can think of a number of ways:

* Append comprehensive `inspector` type definitions to the end of `types/node/index.d.ts` (what this PR did when I first opened it)
* Add comprehensive `inspector` type definitions in a different `d.ts` file, and link it in `types/node/index.d.ts` (exactly what this PR is doing currently)
* Add basic type definitions (no auto-generated interfaces) here, and leave more comprehensive type definitions for a `@types/inspector` module (again, not sure if this is possible, since definitions may clash if someone installs both `@types/node` and `@types/inspector`)
* Add comprehensive type definitions as `@types/inspector`, and leave `@types/node` alone

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  * https://nodejs.org/dist/latest/docs/api/inspector.html
  * https://chromedevtools.github.io/devtools-protocol/v8/

### CODE REVIEW

**Comment by wwwy3y3** [comment]
@kjin Nice work! I like the second way that putting `inspector` type definitions in a different d.ts file.

**Comment by kjin** [comment]
@wwwy3y3 Thanks for the feedback!

I found that [`baidumap-web-sdk` types](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/baidumap-web-sdk) splits its types into different files. I'll update this PR branch to follow suit.

**Comment by kjin** [comment]
I believe CI failures are unrelated.

File: types/node/node-tests.ts [hunk 1]
```diff
@@ -3380,3 +3381,45 @@ namespace http2_tests {
         str = constants.HTTP2_METHOD_VERSION_CONTROL;
     }
 }
+
+///////////////////////////////////////////////////////////
+/// HTTP/2 Tests                                        ///
```
Comment (wwwy3y3) [inline types/node/node-tests.ts]: should be inspector tests

Comment (kjin) [reply types/node/node-tests.ts]: 😨

**Comment by wwwy3y3** [comment]
@kjin  LGTM 👍

**Comment by kjin** [comment]
@andy-ms Would you mind merging this PR if it looks good? I think the failures are unrelated (from comparing failures with another PR for Node types).

**Comment by andy-ms** [comment]
@kjin Could you add the generate script as `types/node/scripts/generate-inspector.ts`?

**Comment by kjin** [comment]
Sure. I'll clean it up a bit and then add it.

**Comment by kjin** [comment]
@andy-ms How would I specify module dependencies for the script?

It has few enough dependencies (trim-right and devtools-protocol) that it's not strictly necessary. But just wondering if there is a precedent.

**Comment by kjin** [comment]
Never mind, I re-implemented `trim-right` and `devtools-protocol` is extraneous. I've added the script in a `generate-inspector` directory along with a README.

File: types/node/inspector.d.ts [hunk 2]
```diff
@@ -0,0 +1,2484 @@
+// Type definitions for inspector (built-in module in Node 8+)
```
Comment (andy-ms) [inline types/node/inspector.d.ts]: This shouldn't have its own header. Just one for the whole package in `index.d.ts`.

File: types/node/inspector.d.ts [hunk 3]
```diff
@@ -0,0 +1,2484 @@
...
+// Project: http://nodejs.org/
+// Definitions by: Kelvin Jin <https://github.com/kjin>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+// Note: These definitions are auto-generated.
+// Please see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19330
+// for more information.
+
+/// <reference path="./index.d.ts" />
```
Comment (andy-ms) [inline types/node/inspector.d.ts]: Shouldn't be necessary if this is referenced from index?

File: types/node/inspector.d.ts [hunk 4]
```diff
@@ -0,0 +1,2484 @@
...
+
+        /**
+         * Immediately close the session. All pending message callbacks will be called with an error. session.connect() will need to be called to be able to send messages again. Reconnected session will lose all inspector state, such as enabled agents or configured breakpoints.
+         */
+        disconnect(): void;
+
+        /**
+         * Posts a message to the inspector back-end. callback will be notified when a response is received. callback is a function that accepts two optional arguments - error and message-specific result.
+         */
+        post(method: string, params?: {}, callback?: (err: Error, params?: {}) => void): void;
```
Comment (andy-ms) [inline types/node/inspector.d.ts]: `err` is probably potentially null / undefined? Same for most callbacks here.

**Comment by kjin** [comment]
Ok, I've addressed comments and rebased.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**